### ISSUE TITLE
block_in_if_condition_stmt: false positive with blocks inside iterator method calls

### ISSUE THREAD

**Issue by birkenfeld** [rust-lang/rust-clippy#1141]
Example:

```
fn main() {
    let mut range = 0..10;
    debug_assert!(range.all(|i| { i < 10 }));
}
```

Maybe just an in_macro check missing?

**Comment by oli-obk** [comment]
actually this would also occur in `if (0..10).all(|i| { i < 10 })`, the reasoning is in the corresponding unit test:

> // this is a sneaky case, where the block isn't directly in the condition, but is actually
    // inside a closure that the condition is using.  same principle applies.  add some extra
    // expressions to make sure linter isn't confused by them.

the "same principle" being

> in an 'if' condition, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a 'let'

The above case should probably be linted against differently (disallow blocks without statements)

**Comment by matthiaskrgr** [comment]
Just ran into this
````rust
let v = vec![1,2,3];

if v.into_iter().any(|x| {x == 4 }) {
    println!("contains 4!");
}
````

```
warning: in an `if` condition, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
 --> src/main.rs:6:26
  |
6 | if v.into_iter().any(|x| {x == 4 }) {
  |                          ^^^^^^^^^
  |
  = note: `#[warn(clippy::block_in_if_condition_stmt)]` on by default
```

IMO we should not lint here..
